cpp/serve/sampler/cpu_sampler.cc

```diff
@@ -294,8 +295,7 @@ class CPUSampler : public SamplerObj {
           RECORD_EVENT(this->trace_recorder_, request_ids[i], "start sample token");
           // Sample top p from probability.
           sample_results[i].sampled_token_id = SampleTopPFromProb(
-              probs_host, sample_indices[i],
-              generation_cfg[i]->temperature < eps_ ? 0.0 : generation_cfg[i]->top_p,
+              probs_host, i, generation_cfg[i]->temperature < eps_ ? 0.0 : generation_cfg[i]->top_p,
```

Comment:
Is this line changed by mistake?

Reply:
No. We should use `i` to index output probs. Otherwise, `sample_indices[i]` may have repeated values, the parallel sampling may have read-write conflicts.

Reply:
In the sampler, when generating in parallel, `sample_indices` may be [0, 0, 0, 1, 1, 1]. In this way, the problem lies in the writing of `output_prob_dist`. Different threads will write at the same position of `output_prob_dist` at the same time, such as three threads write to position 0 of `output_prob_dist,` and three threads write to position 1. I think that it actually requires six threads to write to different positions of `output_prob_dist`. There is no such problem in the previous prefill because the `BatchSampleTokens` used there do not need to record `output_prob_dist`. The `sample_indices` in decode and draft are continuous ranges, and there will be no conflicts. However, in eagle, prefill needs to record `output_prob_dist`, and conflicts will appear.